### PR TITLE
MDCT-1910: Ratio Input

### PR DESCRIPTION
**Pull request by ntsummers1**
## Description
Fixes [MDCT-1920](https://qmacbis.atlassian.net/browse/MDCT-1910)

https://user-images.githubusercontent.com/19439679/188916823-3cab3ee4-3127-4ef8-bbd3-376b89166a1a.mov

### How to test
./dev local
Create a program
Checkout the temporary ratio mask input field on the first program page

## Code author checklist
- [ ] I have performed a self-review of my code
- [x] I have added [thorough](https://bit.ly/3zPrxuZ) tests
- [x] I have added analytics, if necessary
- [x] I have updated the documentation, if necessary

## Reviewer checklist (two different people)
- [ ] I have done the deep review and verified the items in the above checklist are g2g
- [ ] I have done the lgtm review

### CODE REVIEW

File: services/ui-src/src/utils/forms/schemas.ts
Comment (ntsummers1) [inline services/ui-src/src/utils/forms/schemas.ts:48]: Would love thoughts on how to make this more readable

**Comment by ntsummers1** [comment]
Would love thoughts on the schemas and mask file about how readable that code is. I've added comments and tried to reduce the code down but it still might be a little complex

File: services/ui-src/src/components/fields/NumberField.test.tsx
Comment (gmrabian) [inline services/ui-src/src/components/fields/NumberField.test.tsx:152]: Do we want this to be the case? A single colon is listed in the "bad test cases" below

Comment (gmrabian) [reply services/ui-src/src/components/fields/NumberField.test.tsx:152]: I guess maybe it's ok to leave it, it'll just fail validation, right?

File: services/ui-src/src/forms/mcpar/mcpar.json
Comment (gmrabian) [inline services/ui-src/src/forms/mcpar/mcpar.json:31]: Just making this note so we remember to remove this before merge

File: services/ui-src/src/forms/mcpar/mcpar.schema.ts
Comment (gmrabian) [inline services/ui-src/src/forms/mcpar/mcpar.schema.ts:20]: same as above - remove before merge

File: services/ui-src/src/utils/forms/schemas.test.tsx
Comment (gmrabian) [inline services/ui-src/src/utils/forms/schemas.test.tsx:56]: This is the same as the `testNumberSchema` method above. Do we need both, and both tests? Could we not just do one?

File: services/ui-src/src/utils/forms/schemas.ts
Comment (gmrabian) [inline services/ui-src/src/utils/forms/schemas.ts:18]: Is it possible to make a separate ratio field that validates against number : number ? I feel like that'd be pretty smooth

File: services/ui-src/src/components/fields/NumberField.test.tsx
Comment (ntsummers1) [reply services/ui-src/src/components/fields/NumberField.test.tsx:152]: Right, that'll fail validation so it wouldn't get through on the form. Basically this test is to make sure the masking gets called and that mask returns the right value back when you tab away.

File: services/ui-src/src/utils/forms/schemas.test.tsx
Comment (ntsummers1) [reply services/ui-src/src/utils/forms/schemas.test.tsx]: They were slightly different, however I reduced it down to the same function now as I agree that they are too similar.

File: services/ui-src/src/utils/forms/schemas.ts
Comment (ntsummers1) [reply services/ui-src/src/utils/forms/schemas.ts]: Separated out!

Comment (gmrabian) [reply services/ui-src/src/utils/forms/schemas.ts]: I'm glad this is separated, though my idea was wondering if it's possible to do like
`ratio: mixed(number(), ":", number())` if that makes sense. Like having the schema validate there are two numbers with a colon. Maybe there's no easy way to do that except for regex. Just curious since that would simplify it a lot!

Comment (ntsummers1) [reply services/ui-src/src/utils/forms/schemas.ts]: Not to my knowledge, no :( .. And if i try to do something like that then I'd come back to the complexity before I separated it out.

File: services/ui-src/src/forms/mcpar/mcpar.json
Comment (braxex) [reply services/ui-src/src/forms/mcpar/mcpar.json:31]: +1

File: services/ui-src/src/forms/mcpar/mcpar.schema.ts
Comment (braxex) [inline services/ui-src/src/forms/mcpar/mcpar.schema.ts:21]: remove before merge

File: services/ui-src/src/utils/forms/schemas.ts
Comment (braxex) [inline services/ui-src/src/utils/forms/schemas.ts:76]: Required message should fire when the field is empty instead of the ratio message. Just need to move these two above the first test. Might need to make same change for Number.

File: services/ui-src/src/utils/other/mask.test.ts
Comment (braxex) [inline services/ui-src/src/utils/other/mask.test.ts:96]: If we're gonna go here, might as well go all in. 😉 
```suggestion
      { test: "No numbers here, m'lord", expected: ":" },
```

**Review by braxex — APPROVED** [review]
One change needed in the order of tests for ratio validation, but otherwise approved once that's done. Remember to take out the test cases before merging.

**Review by gmrabian — APPROVED** [review]
works nicely locally

File: services/ui-src/src/utils/other/mask.test.ts
Comment (ntsummers1) [reply services/ui-src/src/utils/other/mask.test.ts]: Fixed :)

File: services/ui-src/src/utils/forms/schemas.ts
Comment (ntsummers1) [reply services/ui-src/src/utils/forms/schemas.ts]: Worked like a charm. Fixed!

**Pull request merged**
